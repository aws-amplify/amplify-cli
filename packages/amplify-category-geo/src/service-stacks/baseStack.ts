import * as cdk from '@aws-cdk/core';
import { prepareApp } from '@aws-cdk/core/lib/private/prepare-app';
import { $TSObject } from 'amplify-cli-core';

export type TemplateMappings = {
  RegionMapping: $TSObject;
};

export class BaseStack extends cdk.Stack {
  protected parameters: Map<string, cdk.CfnParameter>;
  protected regionMapping: cdk.CfnMapping;

  constructor(scope: cdk.Construct, id: string, props: TemplateMappings) {
    super(scope, id);
    this.parameters = new Map();

    this.regionMapping = new cdk.CfnMapping(this, 'RegionMapping', {
      mapping: props.RegionMapping,
    });
  }

  // construct the stack CFN input parameters
  constructInputParameters(parameterNames: Array<string>): Map<string, cdk.CfnParameter> {
    let parametersMap: Map<string, cdk.CfnParameter> = new Map();
    parameterNames.forEach(parameterName => {
      const inputParameter = new cdk.CfnParameter(this, parameterName, { type: 'String' });
      parametersMap.set(parameterName, inputParameter);
    });
    return parametersMap;
  }

  toCloudFormation() {
    prepareApp(this);
    const cfn = this._toCloudFormation();
    return cfn;
  }
}
