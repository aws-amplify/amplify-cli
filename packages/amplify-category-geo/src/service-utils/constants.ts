import * as path from 'path';

export const apiDocs = {
  mapStyles: 'https://docs.aws.amazon.com/location-maps/latest/APIReference/API_MapConfiguration.html',
  pricingPlan: 'https://aws.amazon.com/location/pricing/',
  dataSourceUsage: 'https://docs.aws.amazon.com/location-places/latest/APIReference/API_DataSourceConfiguration.html',
  locationServiceTerms: 'https://aws.amazon.com/service-terms/'
};

export const previewBanner = 'Amplify Geo category is in developer preview and not intended for production use at this time.';
export const chooseServiceMessageAdd = 'Select which capability you want to add:';
export const chooseServiceMessageUpdate = 'Select which capability you want to update:';
export const chooseServiceMessageRemove = 'Select which capability you want to remove:';
export const parametersFileName = 'parameters.json';
export const provider = 'awscloudformation';
export const customMapLambdaCodePath = path.join(__dirname, '../../resources/custom-map-resource-handler.js');
export const customPlaceIndexLambdaCodePath = path.join(__dirname, '../../resources/custom-place-index-resource-handler.js');

export enum ServiceName {
  Map = 'Map',
  PlaceIndex = 'PlaceIndex',
  GeofenceCollection = 'GeofenceCollection',
  Tracker = 'Tracker',
}
