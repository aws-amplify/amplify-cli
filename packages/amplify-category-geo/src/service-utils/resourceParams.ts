import { ProviderContext } from 'amplify-cli-core';

/**
 * Parameters common to geo resources
 */
export type ResourceParameters = {
    providerContext: ProviderContext,
    name: string,
    accessType: AccessType,
    isDefault: boolean,
    dataProvider: DataProvider
}

export enum AccessType {
    AuthorizedUsers = 'AuthorizedUsers',
    AuthorizedAndGuestUsers = 'AuthorizedAndGuestUsers'
}

export enum DataProvider {
    Esri = 'Esri',
    Here = 'HERE'
}
