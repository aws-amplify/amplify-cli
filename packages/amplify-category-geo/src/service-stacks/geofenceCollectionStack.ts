import * as cdk from '@aws-cdk/core';
import * as iam from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';
import { Duration, Fn } from '@aws-cdk/core';
import { GeofenceCollectionParameters } from '../service-utils/geofenceCollectionParams';
import { BaseStack, TemplateMappings } from './baseStack';
import { Effect } from '@aws-cdk/aws-iam';
import { customGeofenceCollectionLambdaCodePath } from '../service-utils/constants';
import * as fs from 'fs-extra';
import _ from 'lodash';
import { Runtime } from '@aws-cdk/aws-lambda';
import { crudPermissionsMap } from '../service-utils/geofenceCollectionUtils';

type GeofenceCollectionStackProps = Pick<GeofenceCollectionParameters, 'groupPermissions'> &
    TemplateMappings & { authResourceName: string };

export class GeofenceCollectionStack extends BaseStack {
  protected readonly groupPermissions: Record<string, string[]>;
  protected readonly geofenceCollectionResource: cdk.CustomResource;
  protected readonly geofenceCollectionRegion: string;
  protected readonly geofenceCollectionName: string;
  protected readonly authResourceName: string;

  constructor(scope: cdk.Construct, id: string, private readonly props: GeofenceCollectionStackProps) {
    super(scope, id, props);

    this.groupPermissions = this.props.groupPermissions;
    this.authResourceName = this.props.authResourceName;
    this.geofenceCollectionRegion = this.regionMapping.findInMap(cdk.Fn.ref('AWS::Region'), 'locationServiceRegion');

    const inputParameters: string[] = Object.keys(this.props.groupPermissions).map(
        (group: string) => `authuserPoolGroups${group}GroupRole`
    );
    inputParameters.push(
      `auth${this.authResourceName}UserPoolId`,
      'collectionName',
      'env',
      'isDefault'
    );
    this.parameters = this.constructInputParameters(inputParameters);

    this.geofenceCollectionName = Fn.join('-', [this.parameters.get('collectionName')!.valueAsString, this.parameters.get('env')!.valueAsString]);

    this.geofenceCollectionResource = this.constructCollectionResource();
    this.constructCollectionPolicyResources(this.geofenceCollectionResource);
    this.constructOutputs();
  }

  private constructOutputs() {
    new cdk.CfnOutput(this, 'Name', {
      value: this.geofenceCollectionResource.getAtt('CollectionName').toString()
    });
    new cdk.CfnOutput(this, 'Region', {
      value: this.geofenceCollectionRegion
    });
    // This is a work-around until the CollectionArn is included in the `UpdateGeofenceCollection` output
    const outputGeofenceCollectionARN = cdk.Fn.sub('arn:aws:geo:${region}:${account}:geofence-collection/${collectionName}', {
      region: this.geofenceCollectionRegion,
      account: cdk.Fn.ref('AWS::AccountId'),
      collectionName: this.geofenceCollectionResource.getAtt('CollectionName').toString()
    });
    new cdk.CfnOutput(this, 'Arn', {
      // value: this.geofenceCollectionResource.getAtt('CollectionArn').toString(),
      value: outputGeofenceCollectionARN
    });
  }

  private constructCollectionResource(): cdk.CustomResource {
    const geoCreateCollectionStatement = new iam.PolicyStatement({
      effect: Effect.ALLOW,
    });
    geoCreateCollectionStatement.addActions('geo:CreateGeofenceCollection');
    geoCreateCollectionStatement.addAllResources();

    const geofenceCollectionARN = cdk.Fn.sub('arn:aws:geo:${region}:${account}:geofence-collection/${collectionName}', {
      region: this.geofenceCollectionRegion,
      account: cdk.Fn.ref('AWS::AccountId'),
      collectionName: this.geofenceCollectionName,
    });

    const geoUpdateDeleteCollectionStatement = new iam.PolicyStatement({
      effect: Effect.ALLOW,
    });
    geoUpdateDeleteCollectionStatement.addActions('geo:UpdateGeofenceCollection', 'geo:DeleteGeofenceCollection');
    geoUpdateDeleteCollectionStatement.addResources(geofenceCollectionARN);

    const customGeofenceCollectionLambdaCode = fs.readFileSync(customGeofenceCollectionLambdaCodePath, 'utf-8');
    const customGeofenceCollectionLambda = new lambda.Function(this, 'CustomGeofenceCollectionLambda', {
      code: lambda.Code.fromInline(customGeofenceCollectionLambdaCode),
      handler: 'index.handler',
      runtime: Runtime.NODEJS_14_X,
      timeout: Duration.seconds(300),
    });
    customGeofenceCollectionLambda.addToRolePolicy(geoCreateCollectionStatement);
    customGeofenceCollectionLambda.addToRolePolicy(geoUpdateDeleteCollectionStatement);

    const geofenceCollectionCustomResource = new cdk.CustomResource(this, 'CustomGeofenceCollection', {
      serviceToken: customGeofenceCollectionLambda.functionArn,
      resourceType: 'Custom::LambdaCallout',
      properties: {
        collectionName: this.geofenceCollectionName,
        region: this.geofenceCollectionRegion,
        env: cdk.Fn.ref('env')
      },
    });

    return geofenceCollectionCustomResource;
  }

  // Grant selected access permissions for Geofence operations to chosen Cognito Groups
  private constructCollectionPolicyResources(collectionResource: cdk.CustomResource) {
    Object.keys(this.groupPermissions).forEach((group: string) => {
      // This is a work-around until the CollectionArn is included in the `UpdateGeofenceCollection` output
      const outputGeofenceCollectionARN = cdk.Fn.sub('arn:aws:geo:${region}:${account}:geofence-collection/${collectionName}', {
        region: this.geofenceCollectionRegion,
        account: cdk.Fn.ref('AWS::AccountId'),
        collectionName: collectionResource.getAtt('CollectionName').toString()
      });
      const crudActions: string[] = _.uniq(_.flatten(this.groupPermissions[group].map((permission: string) => crudPermissionsMap[permission])));
      const policyDocument = new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: crudActions,
              resources: [outputGeofenceCollectionARN],
            }),
          ],
      });

      const policy = new iam.CfnPolicy(this, `${group}GeofenceCollectionPolicy`, {
          policyName: `${group}${this.geofenceCollectionName}Policy`,
          roles: [
              cdk.Fn.join('-',
              [
                this.parameters.get(`auth${this.authResourceName}UserPoolId`)!.valueAsString,
                `${group}GroupRole`
              ])
          ],
          policyDocument: policyDocument,
      });
    });
  }
}
