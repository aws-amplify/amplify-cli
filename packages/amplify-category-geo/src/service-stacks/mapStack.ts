import * as cdk from '@aws-cdk/core';
import * as iam from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';
import { MapParameters } from '../service-utils/mapParams';
import { CfnResource, Duration, Fn } from '@aws-cdk/core';
import { AccessType } from '../service-utils/resourceParams';
import { BaseStack, TemplateMappings } from './baseStack';
import { Effect } from '@aws-cdk/aws-iam';
import { Runtime } from '@aws-cdk/aws-lambda';
import { customMapLambdaCodePath } from '../service-utils/constants';
import * as fs from 'fs-extra';

type MapStackProps = Pick<MapParameters, 'accessType'> & TemplateMappings;

export class MapStack extends BaseStack {
  protected readonly accessType: string;
  protected readonly mapResource: cdk.CustomResource;
  protected readonly mapRegion: string;
  protected readonly mapName: string;

  constructor(scope: cdk.Construct, id: string, private readonly props: MapStackProps) {
    super(scope, id, props);

    this.accessType = this.props.accessType;
    this.mapRegion = this.regionMapping.findInMap(cdk.Fn.ref('AWS::Region'), 'locationServiceRegion');

    this.parameters = this.constructInputParameters([
      'authRoleName',
      'unauthRoleName',
      'mapName',
      'mapStyle',
      'env',
      'isDefault',
    ]);

    this.mapName = Fn.join('-', [this.parameters.get('mapName')!.valueAsString, this.parameters.get('env')!.valueAsString]);
    this.mapResource = this.constructMapResource();
    this.constructMapPolicyResource(this.mapResource);
    this.constructOutputs();
  }

  private constructOutputs() {
    new cdk.CfnOutput(this, 'Name', {
      value: this.mapResource.getAtt('MapName').toString(),
    });
    new cdk.CfnOutput(this, 'Style', {
      value: this.parameters.get('mapStyle')!.valueAsString,
    });
    new cdk.CfnOutput(this, 'Region', {
      value: this.mapRegion,
    });
    new cdk.CfnOutput(this, 'Arn', {
      value: this.mapResource.getAtt('MapArn').toString(),
    });
  }

  private constructMapResource(): cdk.CustomResource {
    const geoCreateMapStatement = new iam.PolicyStatement({
      effect: Effect.ALLOW,
    });
    geoCreateMapStatement.addActions('geo:CreateMap');
    geoCreateMapStatement.addAllResources();

    const mapARN = cdk.Fn.sub('arn:aws:geo:${region}:${account}:map/${mapName}', {
      region: this.mapRegion,
      account: cdk.Fn.ref('AWS::AccountId'),
      mapName: this.mapName,
    });

    const geoUpdateDeleteMapStatement = new iam.PolicyStatement({
      effect: Effect.ALLOW,
    });
    geoUpdateDeleteMapStatement.addActions('geo:UpdateMap', 'geo:DeleteMap');
    geoUpdateDeleteMapStatement.addResources(mapARN);

    const mapStyle = this.parameters.get('mapStyle')!.valueAsString;

    const customMapLambdaCode = fs.readFileSync(customMapLambdaCodePath, 'utf-8');
    const customMapLambda = new lambda.Function(this, 'CustomMapLambda', {
      code: lambda.Code.fromInline(customMapLambdaCode),
      handler: 'index.handler',
      runtime: Runtime.NODEJS_14_X,
      timeout: Duration.seconds(300),
    });
    customMapLambda.addToRolePolicy(geoCreateMapStatement);
    customMapLambda.addToRolePolicy(geoUpdateDeleteMapStatement);

    const mapCustomResource = new cdk.CustomResource(this, 'CustomMap', {
      serviceToken: customMapLambda.functionArn,
      resourceType: 'Custom::LambdaCallout',
      properties: {
        mapName: this.mapName,
        mapStyle: mapStyle,
        region: this.mapRegion,
        env: cdk.Fn.ref('env'),
      },
    });

    return mapCustomResource;
  }

  // Grant read-only access to the Map for Authorized and/or Guest users
  private constructMapPolicyResource(mapResource: cdk.CustomResource): CfnResource {
    let policy = new iam.PolicyDocument({
      statements: [
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: ['geo:GetMapStyleDescriptor', 'geo:GetMapGlyphs', 'geo:GetMapSprites', 'geo:GetMapTile'],
          resources: [mapResource.getAtt('MapArn').toString()],
        }),
      ],
    });

    let cognitoRoles: Array<string> = new Array();
    cognitoRoles.push(this.parameters.get('authRoleName')!.valueAsString);
    if (this.accessType == AccessType.AuthorizedAndGuestUsers) {
      cognitoRoles.push(this.parameters.get('unauthRoleName')!.valueAsString);
    }

    return new iam.CfnPolicy(this, 'MapPolicy', {
      policyName: `${this.mapName}Policy`,
      roles: cognitoRoles,
      policyDocument: policy,
    });
  }
}
