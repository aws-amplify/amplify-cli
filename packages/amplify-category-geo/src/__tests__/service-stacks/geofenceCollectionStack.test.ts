import { AccessType } from '../../service-utils/resourceParams';
import { GeofenceCollectionStack } from '../../service-stacks/geofenceCollectionStack';
import { App } from '@aws-cdk/core';

describe('cdk stack creation for geofence collection service', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('does not add geofence collection policy for empty group permissions', async () => {
    const stackProps = {
        accessType: AccessType.AuthorizedUsers,
        RegionMapping: {
            'eu-west-2': {
                locationServiceRegion: 'eu-central-1',
            },
        },
        groupPermissions: {},
        authResourceName: 'mockAuthResource123'
    };
    const geofenceCollectionStack = new GeofenceCollectionStack(new App(), 'GeofenceCollectionStack', stackProps);
    expect(geofenceCollectionStack.toCloudFormation()).toMatchSnapshot();
  });

  it('creates geofence collection policy for given group permissions', async () => {
    const stackProps = {
        accessType: AccessType.AuthorizedUsers,
        RegionMapping: {
          'eu-west-2': {
            locationServiceRegion: 'eu-central-1',
          },
        },
        groupPermissions: {
          collectionsAdmin: [
            "Read geofence",
            "Create/Update geofence"
          ]
        },
        authResourceName: 'mockAuthResource123'
    };
    const geofenceCollectionStack = new GeofenceCollectionStack(new App(), 'GeofenceCollectionStack', stackProps);
    expect(geofenceCollectionStack.toCloudFormation()).toMatchSnapshot();
  });
});
