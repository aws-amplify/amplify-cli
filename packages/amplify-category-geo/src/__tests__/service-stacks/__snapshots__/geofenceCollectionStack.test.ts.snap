// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cdk stack creation for geofence collection service creates geofence collection policy for given group permissions 1`] = `
Object {
  "Mappings": Object {
    "RegionMapping": Object {
      "eu-west-2": Object {
        "locationServiceRegion": "eu-central-1",
      },
    },
  },
  "Outputs": Object {
    "Arn": Object {
      "Value": Object {
        "Fn::Sub": Array [
          "arn:aws:geo:\${region}:\${account}:geofence-collection/\${collectionName}",
          Object {
            "account": Object {
              "Ref": "AWS::AccountId",
            },
            "collectionName": Object {
              "Fn::GetAtt": Array [
                "CustomGeofenceCollection",
                "CollectionName",
              ],
            },
            "region": Object {
              "Fn::FindInMap": Array [
                "RegionMapping",
                Object {
                  "Ref": "AWS::Region",
                },
                "locationServiceRegion",
              ],
            },
          },
        ],
      },
    },
    "Name": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "CustomGeofenceCollection",
          "CollectionName",
        ],
      },
    },
    "Region": Object {
      "Value": Object {
        "Fn::FindInMap": Array [
          "RegionMapping",
          Object {
            "Ref": "AWS::Region",
          },
          "locationServiceRegion",
        ],
      },
    },
  },
  "Parameters": Object {
    "authmockAuthResource123UserPoolId": Object {
      "Type": "String",
    },
    "authuserPoolGroupscollectionsAdminGroupRole": Object {
      "Type": "String",
    },
    "collectionName": Object {
      "Type": "String",
    },
    "dataProvider": Object {
      "Type": "String",
    },
    "env": Object {
      "Type": "String",
    },
    "isDefault": Object {
      "Type": "String",
    },
    "pricingPlan": Object {
      "Type": "String",
    },
  },
  "Resources": Object {
    "CustomGeofenceCollection": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomGeofenceCollectionLambdaCA3D002B",
            "Arn",
          ],
        },
        "collectionName": Object {
          "Fn::Join": Array [
            "-",
            Array [
              Object {
                "Ref": "collectionName",
              },
              Object {
                "Ref": "env",
              },
            ],
          ],
        },
        "dataSource": Object {
          "Ref": "dataProvider",
        },
        "env": Object {
          "Ref": "env",
        },
        "pricingPlan": Object {
          "Ref": "pricingPlan",
        },
        "region": Object {
          "Fn::FindInMap": Array [
            "RegionMapping",
            Object {
              "Ref": "AWS::Region",
            },
            "locationServiceRegion",
          ],
        },
      },
      "Type": "Custom::LambdaCallout",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomGeofenceCollectionLambdaCA3D002B": Object {
      "DependsOn": Array [
        "CustomGeofenceCollectionLambdaServiceRoleDefaultPolicy0A18B369",
        "CustomGeofenceCollectionLambdaServiceRole9040D551",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "const response = require('cfn-response');
const aws = require('aws-sdk');
exports.handler = async function (event, context) {
  try {
    console.log('REQUEST RECEIVED:' + JSON.stringify(event));
    if (event.RequestType == 'Create') {
      const params = {
        CollectionName: event.ResourceProperties.collectionName,
        PricingPlan: event.ResourceProperties.pricingPlan
      };
      if (params.PricingPlan !== 'RequestBasedUsage') {
        params['PricingPlanDataSource'] = event.ResourceProperties.dataSource;
      }
      const locationClient = new aws.Location({ apiVersion: '2020-11-19', region: event.ResourceProperties.region });
      const res = await locationClient.createGeofenceCollection(params).promise();
      console.log('create resource response data' + JSON.stringify(res));
      if (res.CollectionName && res.CollectionArn) {
        await response.send(event, context, response.SUCCESS, res, params.CollectionName);
      } else {
        await response.send(event, context, response.FAILED, res, params.CollectionName);
      }
    }
    if (event.RequestType == 'Update') {
      const params = {
        CollectionName: event.ResourceProperties.collectionName,
        PricingPlan: event.ResourceProperties.pricingPlan
      };
      if (params.PricingPlan !== 'RequestBasedUsage') {
        params['PricingPlanDataSource'] = event.ResourceProperties.dataSource;
      }
      const locationClient = new aws.Location({ apiVersion: '2020-11-19', region: event.ResourceProperties.region });
      const res = await locationClient.updateGeofenceCollection(params).promise();
      console.log('update resource response data' + JSON.stringify(res));
      if (res.CollectionName) {
        await response.send(event, context, response.SUCCESS, res, params.CollectionName);
      } else {
        await response.send(event, context, response.FAILED, res, params.CollectionName);
      }
    }
    if (event.RequestType == 'Delete') {
      const params = {
        CollectionName: event.ResourceProperties.collectionName
      };
      const locationClient = new aws.Location({ apiVersion: '2020-11-19', region: event.ResourceProperties.region });
      const res = await locationClient.deleteGeofenceCollection(params).promise();
      console.log('delete resource response data' + JSON.stringify(res));
      await response.send(event, context, response.SUCCESS, res, params.CollectionName);
    }
  } catch (err) {
    console.log(err.stack);
    const res = { Error: err };
    await response.send(event, context, response.FAILED, res, event.ResourceProperties.collectionName);
    throw err;
  }
};
",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomGeofenceCollectionLambdaServiceRole9040D551",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomGeofenceCollectionLambdaServiceRole9040D551": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomGeofenceCollectionLambdaServiceRoleDefaultPolicy0A18B369": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "geo:CreateGeofenceCollection",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "geo:UpdateGeofenceCollection",
                "geo:DeleteGeofenceCollection",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Sub": Array [
                  "arn:aws:geo:\${region}:\${account}:geofence-collection/\${collectionName}",
                  Object {
                    "account": Object {
                      "Ref": "AWS::AccountId",
                    },
                    "collectionName": Object {
                      "Fn::Join": Array [
                        "-",
                        Array [
                          Object {
                            "Ref": "collectionName",
                          },
                          Object {
                            "Ref": "env",
                          },
                        ],
                      ],
                    },
                    "region": Object {
                      "Fn::FindInMap": Array [
                        "RegionMapping",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        "locationServiceRegion",
                      ],
                    },
                  },
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomGeofenceCollectionLambdaServiceRoleDefaultPolicy0A18B369",
        "Roles": Array [
          Object {
            "Ref": "CustomGeofenceCollectionLambdaServiceRole9040D551",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "collectionsAdminGeofenceCollectionPolicy": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "geo:GetGeofence",
                "geo:PutGeofence",
                "geo:BatchPutGeofence",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Sub": Array [
                  "arn:aws:geo:\${region}:\${account}:geofence-collection/\${collectionName}",
                  Object {
                    "account": Object {
                      "Ref": "AWS::AccountId",
                    },
                    "collectionName": Object {
                      "Fn::GetAtt": Array [
                        "CustomGeofenceCollection",
                        "CollectionName",
                      ],
                    },
                    "region": Object {
                      "Fn::FindInMap": Array [
                        "RegionMapping",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        "locationServiceRegion",
                      ],
                    },
                  },
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "collectionsAdmin",
              Object {
                "Fn::Join": Array [
                  "-",
                  Array [
                    Object {
                      "Ref": "collectionName",
                    },
                    Object {
                      "Ref": "env",
                    },
                  ],
                ],
              },
              "Policy",
            ],
          ],
        },
        "Roles": Array [
          Object {
            "Fn::Join": Array [
              "-",
              Array [
                Object {
                  "Ref": "authmockAuthResource123UserPoolId",
                },
                "collectionsAdminGroupRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;

exports[`cdk stack creation for geofence collection service does not add geofence collection policy for empty group permissions 1`] = `
Object {
  "Mappings": Object {
    "RegionMapping": Object {
      "eu-west-2": Object {
        "locationServiceRegion": "eu-central-1",
      },
    },
  },
  "Outputs": Object {
    "Arn": Object {
      "Value": Object {
        "Fn::Sub": Array [
          "arn:aws:geo:\${region}:\${account}:geofence-collection/\${collectionName}",
          Object {
            "account": Object {
              "Ref": "AWS::AccountId",
            },
            "collectionName": Object {
              "Fn::GetAtt": Array [
                "CustomGeofenceCollection",
                "CollectionName",
              ],
            },
            "region": Object {
              "Fn::FindInMap": Array [
                "RegionMapping",
                Object {
                  "Ref": "AWS::Region",
                },
                "locationServiceRegion",
              ],
            },
          },
        ],
      },
    },
    "Name": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "CustomGeofenceCollection",
          "CollectionName",
        ],
      },
    },
    "Region": Object {
      "Value": Object {
        "Fn::FindInMap": Array [
          "RegionMapping",
          Object {
            "Ref": "AWS::Region",
          },
          "locationServiceRegion",
        ],
      },
    },
  },
  "Parameters": Object {
    "authmockAuthResource123UserPoolId": Object {
      "Type": "String",
    },
    "collectionName": Object {
      "Type": "String",
    },
    "dataProvider": Object {
      "Type": "String",
    },
    "env": Object {
      "Type": "String",
    },
    "isDefault": Object {
      "Type": "String",
    },
    "pricingPlan": Object {
      "Type": "String",
    },
  },
  "Resources": Object {
    "CustomGeofenceCollection": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomGeofenceCollectionLambdaCA3D002B",
            "Arn",
          ],
        },
        "collectionName": Object {
          "Fn::Join": Array [
            "-",
            Array [
              Object {
                "Ref": "collectionName",
              },
              Object {
                "Ref": "env",
              },
            ],
          ],
        },
        "dataSource": Object {
          "Ref": "dataProvider",
        },
        "env": Object {
          "Ref": "env",
        },
        "pricingPlan": Object {
          "Ref": "pricingPlan",
        },
        "region": Object {
          "Fn::FindInMap": Array [
            "RegionMapping",
            Object {
              "Ref": "AWS::Region",
            },
            "locationServiceRegion",
          ],
        },
      },
      "Type": "Custom::LambdaCallout",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomGeofenceCollectionLambdaCA3D002B": Object {
      "DependsOn": Array [
        "CustomGeofenceCollectionLambdaServiceRoleDefaultPolicy0A18B369",
        "CustomGeofenceCollectionLambdaServiceRole9040D551",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "const response = require('cfn-response');
const aws = require('aws-sdk');
exports.handler = async function (event, context) {
  try {
    console.log('REQUEST RECEIVED:' + JSON.stringify(event));
    if (event.RequestType == 'Create') {
      const params = {
        CollectionName: event.ResourceProperties.collectionName,
        PricingPlan: event.ResourceProperties.pricingPlan
      };
      if (params.PricingPlan !== 'RequestBasedUsage') {
        params['PricingPlanDataSource'] = event.ResourceProperties.dataSource;
      }
      const locationClient = new aws.Location({ apiVersion: '2020-11-19', region: event.ResourceProperties.region });
      const res = await locationClient.createGeofenceCollection(params).promise();
      console.log('create resource response data' + JSON.stringify(res));
      if (res.CollectionName && res.CollectionArn) {
        await response.send(event, context, response.SUCCESS, res, params.CollectionName);
      } else {
        await response.send(event, context, response.FAILED, res, params.CollectionName);
      }
    }
    if (event.RequestType == 'Update') {
      const params = {
        CollectionName: event.ResourceProperties.collectionName,
        PricingPlan: event.ResourceProperties.pricingPlan
      };
      if (params.PricingPlan !== 'RequestBasedUsage') {
        params['PricingPlanDataSource'] = event.ResourceProperties.dataSource;
      }
      const locationClient = new aws.Location({ apiVersion: '2020-11-19', region: event.ResourceProperties.region });
      const res = await locationClient.updateGeofenceCollection(params).promise();
      console.log('update resource response data' + JSON.stringify(res));
      if (res.CollectionName) {
        await response.send(event, context, response.SUCCESS, res, params.CollectionName);
      } else {
        await response.send(event, context, response.FAILED, res, params.CollectionName);
      }
    }
    if (event.RequestType == 'Delete') {
      const params = {
        CollectionName: event.ResourceProperties.collectionName
      };
      const locationClient = new aws.Location({ apiVersion: '2020-11-19', region: event.ResourceProperties.region });
      const res = await locationClient.deleteGeofenceCollection(params).promise();
      console.log('delete resource response data' + JSON.stringify(res));
      await response.send(event, context, response.SUCCESS, res, params.CollectionName);
    }
  } catch (err) {
    console.log(err.stack);
    const res = { Error: err };
    await response.send(event, context, response.FAILED, res, event.ResourceProperties.collectionName);
    throw err;
  }
};
",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomGeofenceCollectionLambdaServiceRole9040D551",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomGeofenceCollectionLambdaServiceRole9040D551": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomGeofenceCollectionLambdaServiceRoleDefaultPolicy0A18B369": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "geo:CreateGeofenceCollection",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "geo:UpdateGeofenceCollection",
                "geo:DeleteGeofenceCollection",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Sub": Array [
                  "arn:aws:geo:\${region}:\${account}:geofence-collection/\${collectionName}",
                  Object {
                    "account": Object {
                      "Ref": "AWS::AccountId",
                    },
                    "collectionName": Object {
                      "Fn::Join": Array [
                        "-",
                        Array [
                          Object {
                            "Ref": "collectionName",
                          },
                          Object {
                            "Ref": "env",
                          },
                        ],
                      ],
                    },
                    "region": Object {
                      "Fn::FindInMap": Array [
                        "RegionMapping",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        "locationServiceRegion",
                      ],
                    },
                  },
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomGeofenceCollectionLambdaServiceRoleDefaultPolicy0A18B369",
        "Roles": Array [
          Object {
            "Ref": "CustomGeofenceCollectionLambdaServiceRole9040D551",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
